src/Common/src/TypeSystem/TypesDebugInfoWriter/UserDefinedTypeDescriptor.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Reflection.Metadata;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeSystem.TypesDebugInfo
+{
+    public class UserDefinedTypeDescriptor
+    {
+        public UserDefinedTypeDescriptor(ITypesDebugInfoWriter objectWriter)
+        {
+            _objectWriter = objectWriter;
+        }
+        public uint GetTypeIndex(TypeDesc type)
+        {
+            uint variableTypeIndex = 0;
+            if (!_knownTypes.TryGetValue(type, out variableTypeIndex))
+            {
+                variableTypeIndex = GetNewTypeIndex(type);
+            }
+            return variableTypeIndex;          
+        }
+
+        private uint GetNewTypeIndex(TypeDesc type)
+        {
+            switch (type.Category)
+            {
+                case TypeFlags.ValueType:
+                case TypeFlags.Class:
+                    return GetClassTypeIndex(type);
+                case TypeFlags.Enum:
+                    return GetEnumTypeIndex(type);
+            }
+            return 0;
+        }
+
+        public uint GetEnumTypeIndex(TypeDesc type)
+        {
+            DefType defType = type as DefType;
+            if (defType == null)
```

Comment:
If someone calls this with wrong parameters, we might want to assert instead of silently generating bad debugging data. It's a bit easier to troubleshoot...